File: packages/python/lizard/platform/projects.py
from __future__ import annotations
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .client import PlatformClient


@dataclass
class Project:
    id: str
    name: str
    slug: str
    workspace_id: str

    @classmethod
    def _from_dict(cls, d: dict) -> "Project":
        return cls(id=d["id"], name=d["name"], slug=d.get("slug", ""), workspace_id=d.get("workspaceId", ""))


class ProjectsAPI:
    def __init__(self, client: "PlatformClient") -> None:
        self._client = client

    def list(self, *, workspace_id: str | None = None) -> list[Project]:
        """List all projects the API key can access."""
        qs = f"?workspaceId={workspace_id}" if workspace_id else ""
        return [Project._from_dict(p) for p in self._client.get(f"/api/projects{qs}")]

    def get(self, id: str) -> Project:
        """Get a project by ID."""
        return Project._from_dict(self._client.get(f"/api/projects/{id}"))

    def create(self, *, workspace_id: str, name: str) -> Project:
        """Create a new project."""
        return Project._from_dict(self._client.post("/api/projects", {"workspaceId": workspace_id, "name": name}))

    def update(self, id: str, *, name: str | None = None) -> Project:
        """Rename a project."""
        return Project._from_dict(self._client.patch(f"/api/projects/{id}", {"name": name}))

    def delete(self, id: str) -> None:
        """Delete a project."""
        self._client.delete(f"/api/projects/{id}")

File: packages/js/src/platform/projects.ts
import type { PlatformClient } from './client'

export interface Project {
  id: string
  name: string
  slug: string
  workspaceId: string
  createdAt?: string
}

export interface CreateProjectOpts {
  workspaceId: string
  name: string
}

export class ProjectsAPI {
  constructor(private readonly client: PlatformClient) {}

  /** List all projects the API key can access. */
  list(opts?: { workspaceId?: string }): Promise<Project[]> {
    const qs = opts?.workspaceId ? `?workspaceId=${opts.workspaceId}` : ''
    return this.client.get(`/api/projects${qs}`)
  }

  /** Get a project by ID. */
  get(id: string): Promise<Project> {
    return this.client.get(`/api/projects/${id}`)
  }

  /** Create a new project. */
  create(opts: CreateProjectOpts): Promise<Project> {
    return this.client.post('/api/projects', opts)
  }

  /** Update a project's name. */
  update(id: string, opts: { name?: string }): Promise<Project> {
    return this.client.patch(`/api/projects/${id}`, opts)
  }

  /** Delete a project. */
  delete(id: string): Promise<void> {
    return this.client.delete(`/api/projects/${id}`)
  }
}
